src/games/strategy/triplea/ai/proAI/ProAI.java

```diff
@@ -482,11 +490,11 @@ public boolean selectAttackSubs(final Territory unitTerritory) {
     final List<Unit> defenders = (List<Unit>) battle.getDefendingUnits();
     ProLogger.info(player.getName() + " checking sub attack in " + unitTerritory + ", attackers=" + attackers
         + ", defenders=" + defenders);
-    ProBattleUtils.setData(getGameData());
+    calc.setData(getGameData());
 
     // Calculate battle results
     final ProBattleResult result =
-        ProBattleUtils.calculateBattleResults(player, unitTerritory, attackers, defenders, new HashSet<Unit>(), true);
+        calc.calculateBattleResults(player, unitTerritory, attackers, defenders, new HashSet<Unit>(), true);
```

Comment:
No. It changes the game data by creating mock battles which is why you don't want to do it in the real game data and need a copy. But you can create as many mock battles as you want in a single copy. You could try to push to copy down or move it to somewhere else but not sure if that would be better. You still need either a set or clear somewhere to indicate when the actual game data has changed (purchase, move, place, etc) and you need a new updated copy.